File: frontend/mobile/lib/core/services/search_service.dart
/// Search service for comprehensive Islamic content search
/// Integrates with backend semantic search service

import 'package:dio/dio.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';
import '../../features/search/data/models/search_models.dart';

class SearchService {
  final DioClient _dioClient;

  SearchService(this._dioClient);

  /// Perform semantic search across all Islamic content
  Future<SearchResponse> search(SearchRequest request) async {
    try {
      final response = await _dioClient.post(
        ApiEndpoints.searchAll,
        data: request.toJson(),
      );

      return SearchResponse.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Search specifically in Quran
  Future<SearchResponse> searchQuran(String query, {
    int limit = 20,
    double minSimilarity = 0.5,
  }) async {
    try {
      final request = SearchRequest(
        query: query,
        limit: limit,
        minSimilarity: minSimilarity,
        contentTypes: ['quran'],
      );

      final response = await _dioClient.post(
        ApiEndpoints.searchQuran,
        data: request.toJson(),
      );

      return SearchResponse.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Search specifically in Hadith
  Future<SearchResponse> searchHadith(String query, {
    int limit = 20,
    double minSimilarity = 0.5,
    List<AuthenticityGrade>? authenticityGrades,
  }) async {
    try {
      final filters = authenticityGrades != null
          ? SearchFilters(authenticityGrades: authenticityGrades)
          : null;

      final request = SearchRequest(
        query: query,
        limit: limit,
        minSimilarity: minSimilarity,
        contentTypes: [
          'sahih_hadith',
          'hasan_hadith',
          'daif_hadith',
          'mawdu_hadith'
        ],
        filters: filters,
      );

      final response = await _dioClient.post(
        ApiEndpoints.searchHadith,
        data: request.toJson(),
      );

      return SearchResponse.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Search in Fatawa (Islamic rulings)
  Future<SearchResponse> searchFatawa(String query, {
    int limit = 20,
    double minSimilarity = 0.5,
  }) async {
    try {
      final request = SearchRequest(
        query: query,
        limit: limit,
        minSimilarity: minSimilarity,
        contentTypes: ['fiqh_ruling', 'scholar_opinion'],
      );

      final response = await _dioClient.post(
        ApiEndpoints.searchFatawa,
        data: request.toJson(),
      );

      return SearchResponse.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Advanced search with full filter support
  Future<SearchResponse> advancedSearch(SearchRequest request) async {
    try {
      final response = await _dioClient.post(
        ApiEndpoints.searchAdvanced,
        data: request.toJson(),
      );

      return SearchResponse.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get search suggestions based on query
  Future<List<QuerySuggestion>> getSuggestions(String query) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.searchSuggestions,
        queryParameters: {'query': query},
      );

      final suggestions = (response.data as List)
          .map((json) => QuerySuggestion.fromJson(json))
          .toList();

      return suggestions;
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Voice search - converts speech to text and performs search
  Future<SearchResponse> voiceSearch(String audioBase64, {
    List<String>? contentTypes,
    int limit = 20,
  }) async {
    try {
      // First, convert speech to text (assuming backend has this endpoint)
      final transcriptionResponse = await _dioClient.post(
        '/api/speech/transcribe',
        data: {'audio': audioBase64, 'language': 'ar'},
      );

      final query = transcriptionResponse.data['text'] as String;

      // Then perform search with transcribed text
      final request = SearchRequest(
        query: query,
        limit: limit,
        contentTypes: contentTypes,
      );

      return await search(request);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Save a search for later access
  Future<SavedSearch> saveSearch(String query, SearchFilters? filters, {
    String? name,
  }) async {
    try {
      final response = await _dioClient.post(
        '/api/search/saved',
        data: {
          'query': query,
          'filters': filters?.toJson(),
          'name': name,
        },
      );

      return SavedSearch.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get all saved searches
  Future<List<SavedSearch>> getSavedSearches() async {
    try {
      final response = await _dioClient.get('/api/search/saved');

      final searches = (response.data as List)
          .map((json) => SavedSearch.fromJson(json))
          .toList();

      return searches;
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Delete a saved search
  Future<void> deleteSavedSearch(String searchId) async {
    try {
      await _dioClient.delete('/api/search/saved/$searchId');
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Handle Dio errors
  Exception _handleError(DioException error) {
    if (error.response != null) {
      final statusCode = error.response!.statusCode;
      final message = error.response!.data['message'] ?? 'حدث خطأ في البحث';

      switch (statusCode) {
        case 400:
          return Exception('طلب بحث غير صالح: $message');
        case 404:
          return Exception('لم يتم العثور على نتائج');
        case 500:
          return Exception('خطأ في الخادم: $message');
        default:
          return Exception('خطأ في البحث: $message');
      }
    } else if (error.type == DioExceptionType.connectionTimeout ||
        error.type == DioExceptionType.receiveTimeout) {
      return Exception('انتهت مهلة الاتصال. يرجى المحاولة مرة أخرى');
    } else if (error.type == DioExceptionType.connectionError) {
      return Exception('خطأ في الاتصال. يرجى التحقق من اتصال الإنترنت');
    } else {
      return Exception('حدث خطأ غير متوقع: ${error.message}');
    }
  }
}

File: frontend/nextjs-app/src/lib/services/search-service.ts
/**
 * Search service for comprehensive Islamic content search
 * Integrates with backend semantic search service
 */

import { apiClient } from '../api/axios-client';
import type {
  SearchRequest,
  SearchResponse,
  QuerySuggestion,
  SavedSearch,
  AuthenticityGrade,
} from '@/types/search';

export class SearchService {
  /**
   * Perform semantic search across all Islamic content
   */
  static async search(request: SearchRequest): Promise<SearchResponse> {
    try {
      return await apiClient.post<SearchResponse>('/api/search/search', request);
    } catch (error) {
      console.error('Failed to perform search:', error);
      throw new Error('فشل البحث. يرجى المحاولة مرة أخرى');
    }
  }

  /**
   * Search specifically in Quran
   */
  static async searchQuran(
    query: string,
    limit: number = 20,
    minSimilarity: number = 0.5
  ): Promise<SearchResponse> {
    try {
      const request: SearchRequest = {
        query,
        limit,
        min_similarity: minSimilarity,
        content_types: ['quran'],
      };

      return await apiClient.post<SearchResponse>('/api/search/quran', request);
    } catch (error) {
      console.error('Failed to search Quran:', error);
      throw new Error('فشل البحث في القرآن الكريم');
    }
  }

  /**
   * Search specifically in Hadith
   */
  static async searchHadith(
    query: string,
    limit: number = 20,
    minSimilarity: number = 0.5,
    authenticityGrades?: AuthenticityGrade[]
  ): Promise<SearchResponse> {
    try {
      const request: SearchRequest = {
        query,
        limit,
        min_similarity: minSimilarity,
        content_types: [
          'sahih_hadith',
          'hasan_hadith',
          'daif_hadith',
          'mawdu_hadith',
        ],
        filters: authenticityGrades
          ? { authenticity_grades: authenticityGrades }
          : undefined,
      };

      return await apiClient.post<SearchResponse>('/api/search/hadith', request);
    } catch (error) {
      console.error('Failed to search Hadith:', error);
      throw new Error('فشل البحث في الأحاديث');
    }
  }

  /**
   * Search in Fatawa (Islamic rulings)
   */
  static async searchFatawa(
    query: string,
    limit: number = 20,
    minSimilarity: number = 0.5
  ): Promise<SearchResponse> {
    try {
      const request: SearchRequest = {
        query,
        limit,
        min_similarity: minSimilarity,
        content_types: ['fiqh_ruling', 'scholar_opinion'],
      };

      return await apiClient.post<SearchResponse>('/api/search/fatawa', request);
    } catch (error) {
      console.error('Failed to search Fatawa:', error);
      throw new Error('فشل البحث في الفتاوى');
    }
  }

  /**
   * Advanced search with full filter support
   */
  static async advancedSearch(request: SearchRequest): Promise<SearchResponse> {
    try {
      return await apiClient.post<SearchResponse>('/api/search/advanced', request);
    } catch (error) {
      console.error('Failed to perform advanced search:', error);
      throw new Error('فشل البحث المتقدم');
    }
  }

  /**
   * Get search suggestions based on query
   */
  static async getSuggestions(query: string): Promise<QuerySuggestion[]> {
    try {
      return await apiClient.get<QuerySuggestion[]>('/api/search/suggestions', {
        params: { query },
      });
    } catch (error) {
      console.error('Failed to get suggestions:', error);
      return [];
    }
  }

  /**
   * Voice search - converts speech to text and performs search
   */
  static async voiceSearch(
    audioBlob: Blob,
    contentTypes?: string[],
    limit: number = 20
  ): Promise<SearchResponse> {
    try {
      // Convert blob to base64
      const base64Audio = await this.blobToBase64(audioBlob);

      // First, convert speech to text
      const transcriptionResponse = await apiClient.post<{ text: string }>(
        '/api/speech/transcribe',
        {
          audio: base64Audio,
          language: 'ar',
        }
      );

      const query = transcriptionResponse.text;

      // Then perform search with transcribed text
      const request: SearchRequest = {
        query,
        limit,
        content_types: contentTypes,
      };

      return await this.search(request);
    } catch (error) {
      console.error('Failed to perform voice search:', error);
      throw new Error('فشل البحث الصوتي');
    }
  }

  /**
   * Save a search for later access
   */
  static async saveSearch(
    query: string,
    filters?: SearchRequest['filters'],
    name?: string
  ): Promise<SavedSearch> {
    try {
      return await apiClient.post<SavedSearch>('/api/search/saved', {
        query,
        filters,
        name,
      });
    } catch (error) {
      console.error('Failed to save search:', error);
      throw new Error('فشل حفظ البحث');
    }
  }

  /**
   * Get all saved searches
   */
  static async getSavedSearches(): Promise<SavedSearch[]> {
    try {
      return await apiClient.get<SavedSearch[]>('/api/search/saved');
    } catch (error) {
      console.error('Failed to get saved searches:', error);
      throw new Error('فشل تحميل البحثات المحفوظة');
    }
  }

  /**
   * Delete a saved search
   */
  static async deleteSavedSearch(searchId: string): Promise<void> {
    try {
      await apiClient.delete(`/api/search/saved/${searchId}`);
    } catch (error) {
      console.error('Failed to delete saved search:', error);
      throw new Error('فشل حذف البحث');
    }
  }

  /**
   * Helper: Convert Blob to base64
   */
  private static blobToBase64(blob: Blob): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onloadend = () => {
        const base64 = reader.result as string;
        // Remove data URL prefix
        const base64Data = base64.split(',')[1];
        resolve(base64Data);
      };
      reader.onerror = reject;
      reader.readAsDataURL(blob);
    });
  }
}
